Deprecation warning for multi table support

imkumarg opened the pull request:
# Description:

Updated the read functions in kartothek (`read_table.*` and `read_dataset_as.*`) with a check to provide the deprecation warning for the multi table support.

Closes part of #317 
Closes #337

fjetter reviewed (changes requested):
An alternative approach would be to check  on the `ds_factory` object whether or not the dataset to be read is multi tabled.
e.g.

```
if len(ds_factory.tables) > 1:
    warning.warn("Trying to read a dataset with multiple internal tables. This functionality will be removed in the next major release. If you require a multi tabled data format, we recommend to switch to the kartothek Cube functionality <link to docs>", DeprecationWarning)
```

This warning should be issued irregardless of what the user input is. If the dataset itself is built this way, it should raise the warning.

imkumarg commented on kartothek/io/dask/dataframe.py:
```diff
@@ -66,6 +66,12 @@ def read_dataset_as_ddf(
 
         For details on performance, see also `dispatch_by`
     """
+    if table is not SINGLE_TABLE or isinstance(columns, dict):
```
> An alternative approach would be to check on the `ds_factory` object whether or not the dataset to be read is multi tabled.
> e.g.
> 
> ```python
> if len(ds_factory.tables) > 1:
>     warning.warn("Trying to read a dataset with multiple internal tables. This functionality will be removed in the next major release. If you require a multi tabled data format, we recommend to switch to the kartothek Cube functionality <link to docs>", DeprecationWarning)
> ```
> 
> This warning should be issued irregardless of what the user input is. If the dataset itself is built this way, it should raise the warning.

Yes it is good to go with this check, I will incorporate this changes.

imkumarg replied:
> We don't necessarily need to assert for the table name to be `SINGLE_TABLE`. Even if the table name would be `my_own_table_name` we can easily infer this during read and this is no issue.

Noted, Will remove the check.

lr4d replied:
@fjetter [here](https://github.com/JDASoftwareGroup/kartothek/pull/337#issuecomment-675347867) you mentioned 
>  the functions which only work on a single table (e.g. everything in dask.dataframe and something like eager.read_table) which should issue a warning if the requested table is not the SINGLE_TABLE constant.

which is why I added that as a task in the issue page [here](https://github.com/JDASoftwareGroup/kartothek/issues/317#issuecomment-686382330)

Am I mixing things up or does this no longer apply?

fjetter replied:
>  here you mentioned

Yes this is what I wrote. Honestly, I can see reasons for both ways. I also mentioned that we shouldn't spam deprecation warnings without a clear migration path. Deprecations should include a suggestion on how to "remove the deprecation". What will be the replacement of a given deprecated feature? Whether or not we warn if the input is or isn't SINGLE_TABLE depends on what the replacement will be.
For instance, a possible replacement would be "as long as the dataset is single-tabled, ignore user input and just infer from dataset" That's what I suggested in #317. Another possibility could be to issue a warning telling the user to provide SINGLE_TABLE (or nothing at all, etc.)
Depending on what the replacement should be, we can formulate the deprecation warnings.

imkumarg replied:
Okay @fjetter, for this PR i will not raise the warnings for the single table and will do that in the follow up PR.

imkumarg commented on kartothek/io/dask/bag.py:
```diff
@@ -68,6 +69,12 @@ def read_dataset_as_metapartitions_bag(
     -------
     A dask.bag object containing the metapartions.
     """
+    if tables is not None or isinstance(columns, dict):
```
The check is done if user wants to read more than single table. It is made sure that the `tables` must not be more than single table.

We can change the condition to `tables is None and isinstance(columns, dict)`. For this case a warning can be raised. Please suggest over it?

lr4d commented on kartothek/io/eager.py:
```diff
@@ -131,6 +131,14 @@ def read_dataset_as_dataframes(
         load_dataset_metadata=True,
     )
 
+    if len(ds_factory.tables) > 1:
```
This will cause the warning to appear twice as you are also raising the warning in read_dataset_as_metapartitions

imkumarg replied:
Got it @lr4d, I will make necessary changes.

lr4d commented on kartothek/io/iter.py:
```diff
@@ -159,6 +167,14 @@ def read_dataset_as_dataframes__iterator(
             {'core': pd.DataFrame, 'lookup': pd.DataFrame}
         ]
     """
+    if tables is None and isinstance(columns, dict):
```
I think we can remove this and also use `if len(ds_factory.tables) > 1:` here, to make sure we always raise the warning if a user reads a multi-tabled dataset.
(You'll have to define `ds_factory = _ensure_factory(...)` before)

imkumarg replied:
Okay will define the `ds_factory` and make the changes.

lr4d reviewed (commented):
a test for this would also be nice

imkumarg commented:
> a test for this would also be nice

Added test case to check for deprecation warning for multi table support.

lr4d commented:
wrt testing I'd recommend to use `pytest.warns` (see https://docs.pytest.org/en/6.1.1/warnings.html#assertwarnings) to check for the warnings being raised in the appropriate scenarios (multi-table datasets, columns/categoricals dicts), it should suffice to put such a test in `kartothek.io.testing.read` for it to apply to all back-ends

fjetter commented on kartothek/io/dask/bag.py:
```diff
@@ -151,6 +153,14 @@ def read_dataset_as_dataframe_bag(
+        warnings.warn(
+            "Trying to read a dataset with multiple internal tables. This functionality will be removed in the next "
+            "major release. If you require a multi tabled data format, we recommend to switch to the kartothek Cube "
+            "functionality https://kartothek.readthedocs.io/en/stable/index.html",
```
You should point to the appropriate section. If we just point to the docs in general, people will get lost.

imkumarg replied:
Sure I will provide the appropriate section link.

fjetter commented on tests/io/dask/bag/test_read.py:
```diff
@@ -65,3 +65,31 @@ def test_read_dataset_as_dataframes_partition_size(store_factory, metadata_versi
         assert bag.npartitions == 2
+
+
+def test_read_dataset_as_dataframe_bag_multi_table_warning(store_factory, metadata_version):
```
This test doesn't verify the warning, does it?

imkumarg replied:
I am on it to verify it with https://docs.pytest.org/en/stable/warnings.html#warns

imkumarg commented:
Added test case applicable for all backends (dask [`bag`, `dataframe`, `delayed`], eager, iter) to validate the multi table deprecation warning. Also updated the document link in the warning message.

fjetter commented on kartothek/io/dask/bag.py:
```diff
@@ -76,6 +76,15 @@ def read_dataset_as_metapartitions_bag(
+            "Trying to read a dataset with multiple internal tables. This functionality will be removed in the next "
+            "major release. If you require a multi tabled data format, we recommend to switch to the kartothek Cube "
+            "functionality. "
+            "https://kartothek.readthedocs.io/en/stable/guide/cube/cube_design_features.html?highlight=multi%20table#multi-dataset-vs-multi-table",
```
Please just point to https://kartothek.readthedocs.io/en/stable/guide/cube/kartothek_cubes.html

imkumarg replied:
Done.

fjetter commented:
for some reason the CI doesn't seem to run. Maybe because the state of your branch is too old. Can you please rebase and resolve the merge conflicts?

lr4d commented on kartothek/io/testing/read.py:
```diff
@@ -723,3 +723,29 @@ def test_extensiondtype_rountrip(store_factory, bound_load_dataframes):
+        dfs=dfs, store=store_factory, dataset_uuid="dataset_uuid"
+    )
+
+    with pytest.warns(DeprecationWarning):
```
np
```suggestion
    with pytest.warns(DeprecationWarning, match="Trying to read a dataset with multiple internal tables.*"):
```

imkumarg replied:
Done.

lr4d commented:
CI is failing due to black

lr4d commented on tests/io/eager/test_read.py:
```diff
@@ -26,7 +26,8 @@ def output_type(request):
 def _read_table(*args, **kwargs):
     if "tables" in kwargs:
-        kwargs.pop("tables")
+        param_tables = kwargs.pop("tables")
+        kwargs["table"] = param_tables
```
same

imkumarg replied:
I am adding `table` here instead of `tables`. If the `table` is not provided the test case will fail for `read_table` in eager.
Also, I need to write a common test case that would be applicable to all the backends. 

https://github.com/JDASoftwareGroup/kartothek/blob/785ec8a1a54aca13f2e7cb076b970ddd0ebb275e/kartothek/io/eager.py#L300

imkumarg commented on tests/io/dask/delayed/test_read.py:
```diff
@@ -23,7 +23,8 @@ def _load_dataframes(output_type, *args, **kwargs):
     elif output_type == "table":
         if "tables" in kwargs:
-            kwargs.pop("tables")
+            param_tables = kwargs.pop("tables")
+            kwargs["table"] = param_tables
```
A quite similar scenario as explained in below comment.

lr4d replied:
Just see that it's table*s*

imkumarg commented:
> Congrats on your first contribution ;)

Thanks, @lr4d
